catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/authentication/kerberos/FilesystemBackendProxy.java

```diff
@@ -0,0 +1,117 @@
+import org.apache.paimon.utils.Preconditions;
+
+/**
+ * Proxy class for FilesystemCatalog to support kerberos authentication. We can also make
+ * FilesystemCatalog as a generic type and pass it as a parameter to the constructor.
+ */
+public class FilesystemBackendProxy implements MethodInterceptor {
+
+  private FileSystemCatalog target;
+  private final CatalogContext catalogContext;
+  private final String kerberosRealm;
+  private final UserGroupInformation proxyUser;
+  private boolean hasUpdatedFileIO = false;
+
+  public FilesystemBackendProxy(
+      FileSystemCatalog target, CatalogContext catalogContext, String kerberosRealm) {
+    this.target = target;
+    this.catalogContext = catalogContext;
+    this.kerberosRealm = kerberosRealm;
+    try {
+      proxyUser = UserGroupInformation.getCurrentUser();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to get current user", e);
+    }
+  }
+
+  @Override
+  public Object intercept(Object o, Method method, Object[] objects, MethodProxy methodProxy)
+      throws Throwable {
+
+    String proxyKerberosPrincipalName = PrincipalUtils.getCurrentPrincipal().getName();
+    if (!proxyKerberosPrincipalName.contains("@")) {
+      proxyKerberosPrincipalName =
+          String.format("%s@%s", proxyKerberosPrincipalName, kerberosRealm);
+    }
+
+    UserGroupInformation realUser =
+        UserGroupInformation.createProxyUser(proxyKerberosPrincipalName, proxyUser);
+
+    return realUser.doAs(
+        (PrivilegedExceptionAction<Object>)
+            () -> {
+              try {
+                updateFileIO(catalogContext);
+                return methodProxy.invoke(target, objects);
+              } catch (Throwable e) {
+                if (RuntimeException.class.isAssignableFrom(e.getClass())) {
+                  throw (RuntimeException) e;
+                }
+                throw new RuntimeException("Failed to invoke method", e);
+              }
+            });
+  }
+
+  public FileSystemCatalog getProxy() {
+    Enhancer e = new Enhancer();
+    e.setClassLoader(target.getClass().getClassLoader());
+    e.setSuperclass(target.getClass());
+    e.setCallback(this);
+    // FileSystemCatalog does not have a no argument constructor.
+    return (FileSystemCatalog)
+        e.create(
+            new Class[] {FileIO.class, Path.class, Options.class},
+            new Object[] {
+              target.fileIO(), new Path(target.warehouse()), new Options(target.options())
+            });
+  }
+
+  /**
+   * Before invoking the doAs method, a Filesystem client has been created when initializing the
+   * FilesystemCatalog. Therefore, the Filesystem client should be recreated within the doAs method
+   * to ensure the ugi corresponds to the impersonated user.
+   */
+  private void updateFileIO(CatalogContext catalogContext)
+      throws IOException, NoSuchFieldException, IllegalAccessException {
+
+    if (hasUpdatedFileIO) {
```

Comment:
hasUpdatedFileIO means UserA and UserB will share the same client?

Reply:
The purpose is to avoid recreating Filesystem client each time when calling  different methods, but it should not be shared by different users.

i have removed `hasUpdatedFileIO`. And i would like to add cache in another pr to avoid recreating Filesystem client frequently.  WDYT? cc @yuqi1129

Reply:
Yeah, using cache to avoid creating clients frequently is meaningful as long as we can confirm the user name is correct. Different users cannot use the same client because it contains user information. 

> And i would like to add cache in another pr to avoid recreating Filesystem client frequently

I think so.

Reply:
ok.